### PR TITLE
Add slurm support WIP

### PR DESCRIPTION
**Pull request by jedwards4b**

### CODE REVIEW

**Comment by rabernat** [comment]
Thanks @jedwards4b for getting this started! I really support this work, as I personally use pangeo on two slurm-based clusters (Columbia's Habanero and TACC Wrangler). I currently have a hacky solution (see #79), so I would be happy to replace it with a more robust approach such as what you are developing here.

File: pangeo/jobqueue.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import toolz
+
+from distributed.utils import tmpfile, ignoring
+
+logger = logging.getLogger(__name__)
+
+class JobQueue(object):
+    """ Base class to launch Dask to a job queue, this class should not be used directly,
+    use inherited class appropriate for your queueing system eg PBScluster or SLURMcluster
+    """
```
Comment (rabernat) [inline pangeo/jobqueue.py]: This seems like the right approach to me: a generic base class with subclasses for the specific scheduler details. 👍

File: pangeo/jobqueue.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
+from contextlib import contextmanager
+import logging
+import subprocess
+import toolz
+
+from distributed.utils import tmpfile, ignoring
+
+logger = logging.getLogger(__name__)
+
+class JobQueue(object):
```
Comment (jhamman) [inline pangeo/jobqueue.py]: pep8: two blank lines before classes

File: pangeo/jobqueue.py (hunk 1 above)
Comment (jhamman) [inline pangeo/jobqueue.py]: Suggest cleaning up this doc string according to [pep8/pep257]:

```Python
    """Baseclass for launching Dask Distributed Clusters for job queues

    This class should not be used directly, use inherited class appropriate for
    your queueing system (e.g. PBScluster or SLURMcluster).
    """
```

File: pangeo/jobqueue.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    """ Base class to launch Dask to a job queue, this class should not be used directly,
+    use inherited class appropriate for your queueing system eg PBScluster or SLURMcluster
+    """
+    def __init__(self):
+        logger.error("JobQueue class should not be called directly")
```
Comment (jhamman) [inline pangeo/jobqueue.py]: maybe just raise a `NotImplementedError` here. Or just remove this method.

File: pangeo/jobqueue.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    def job_file(self):
+        """ Write job submission script to temporary file """
+        with tmpfile(extension='sh') as fn:
+            with open(fn, 'w') as f:
+                f.write(self.job_script())
+            yield fn
+
+    def start_workers(self, n=1):
+        """ Start workers and point them to our local scheduler """
+        outs = []
```
Comment (jhamman) [inline pangeo/jobqueue.py]: `outs` is never used.

File: pangeo/jobqueue.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        self._call([self._cancelcmd] + list(jobs))
+        for w in workers:
+            with ignoring(KeyError):
+                del self.jobs[w]
+
+    def scale_up(self, n, **kwargs):
+        return self.start_workers(n - len(self.jobs))
+
+    def scale_down(self, workers):
+        if isinstance(workers, dict): # https://github.com/dask/distributed/pull/1659
```
Comment (jhamman) [inline pangeo/jobqueue.py]: I don't think this comment is useful is needed anymore. This feature is not in distributed@master.

File: pangeo/slurm.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                 extra='',
+                 **kwargs):
+        """ Initialize a SLURM Cluster
+
+        Parameters
+        ----------
+        name : str
+            Name of worker jobs. Passed to `#SBATCH -J` option.
+        queue : str
+            Destination queue for each worker job. Passed to `#SBATCH -p` option.
```
Comment (jhamman) [inline pangeo/slurm.py]: pep8: wrap this line at 79 characters (or before).

File: pangeo/tests/test_slurm.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import os
+from time import time, sleep
+
+import pytest
+
+from dask.distributed import Client
+from distributed.deploy.adaptive import Adaptive
```
Comment (jhamman) [inline pangeo/tests/test_slurm.py]: Adaptive is not used

File: pangeo/tests/test_slurm.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    with SLURMCluster(walltime='00:02:00', loop=loop) as cluster:
+        cluster.adapt()
+        with Client(cluster) as client:
+            future = client.submit(lambda x: x + 1, 10)
+            assert future.result(60) == 11
+
+            assert cluster.jobs
+
+            start = time()
+            while len(client.scheduler_info()['workers']) != cluster.config['processes']:
```
Comment (jhamman) [inline pangeo/tests/test_slurm.py]: line length

File: pangeo/tests/test_slurm.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            start = time()
+            while cluster.jobs:
+                sleep(0.100)
+                assert time() < start + 10
+
+
+@pytest.mark.skipif('PBS_ACCOUNT' in os.environ, reason='PBS_ACCOUNT defined')
+def test_errors(loop):
+    with pytest.raises(ValueError) as info:
+        PBSCluster()
```
Comment (jhamman) [inline pangeo/tests/test_slurm.py]: Should this be `SLURMCluster`?

**Comment by mrocklin** [comment]
Instead of raising many small style comments I recommend that we switch to running flake8 in the tests

**Comment by mrocklin** [comment]
In principle this approach seems good to me.  I'm glad to see the amount to which the two solutions are the same.

**Comment by jedwards4b** [comment]
@jhamman thanks for the review, I think I've addressed all of your issues.

**Comment by jhamman** [comment]
@jedwards4b - this looks good. Thanks for putting this together, I'm excited to try it out.

General question for those interested in this sort of thing. Do we have any way to test these in a CI system? I'm guessing the answer is probably no for running PBS or Slurm but it would be good to have some tests that we can run apart from HPC machines.

**Pull request merged**